### ISSUE TITLE
Multithreading and download performance

### ISSUE THREAD

**Issue by nrsyed** [Ben-0-mad/YT-TMS-Finder#12]
@Ben-0-mad Just creating this issue to let you know that this project inspired me to try my own hand at developing a general YouTube download/fingerprinting Python package (https://github.com/nrsyed/youtube-audio-matcher) and figured I'd share! It's only half done at the moment—I opted to split the functionality into two modules, one for channel downloading and one for fingerprinting/matching. The download functionality is done and I'm quite satisfied with it. I have yet to start on the fingerprinting/matching.

Features:

- uses multithreading to concurrently download/convert audio from all videos from any number of channels (page source for all channels is obtained concurrently; videos for each channel are downloaded/converted concurrently)
- imports `youtube-dl` Python package, making it cross-platform (no need for youtube-dl executables), and taps directly into the youtube-dl API so there's no need for inefficient subprocess calls or parsing the output of subprocess terminal commands
- allows filtering of videos by duration (e.g., in case you don't want to download videos longer than or shorter than a certain length) and has flexible options for choosing the start time, end time, or duration of extracted audio
- auto-retries failed video downloads (the maximum number of attempts can be specified by the user, or set to unlimited)
- cleanly extracts video metadata from channel page source, including video title
- is set up as a pip installable Python package for easy install
- has a command line interface that can be used from anywhere
- is made with modular design in mind, so lower-level functions (e.g., downloading a single video or extracting video metadata from a channel's page source) can be used on their own and can be incorporated into other applications

If you find anything in it helpful, feel free to use bits and pieces, or add it as a git submodule, etc.

**Comment by Ben-0-mad** [comment]
@nrsyed this is not an issue so I will close it but thank you for letting me know, so does it do exactly the same as mine? I'm still going to keep working on mine if you don't mind.

**Comment by Ben-0-mad** [comment]
Btw good luck with making the fingerprinting thing. Let me know when you finished it.

**Comment by nrsyed** [comment]
@Ben-0-mad Thanks, and good luck to you as well! Yep, the basic goal of my repo is the same. And absolutely, I wasn't suggesting that you should stop working on your project, just wanted to share my approach since you're aiming to address some of the same challenges.

**Comment by nrsyed** [comment]
@Ben-0-mad Figured I'd provide an update! I ended up taking a deep dive into audio fingerprinting and spent the last few weeks writing/documenting an efficient pipeline and a fingerprinting/matching algorithm from scratch. The sphinx documentation (https://nrsyed.github.io/youtube-audio-matcher) and the repo (https://github.com/nrsyed/youtube-audio-matcher) README are largely complete and explain things in more detail, but the highlights are:

* Uses asyncio coroutines and queues with multithreading (for download) and multiprocessing (for fingerprinting and database operations) to efficiently download files and fingerprint/match them in parallel.
* Splits functionality into three standalone submodules: download, database, and audio/fingerprinting, with a main pip installable package that combines them.
* Has a command line interface for the main module and submodules, as well as a Python API so it can be imported as a Python package.
* Can simultaneously process any number of YT channels/users and local files/folders.
* Uses SQLAlchemy for database operations, allowing database functionality to be handled entirely in Python without writing any SQL. This also means any SQL backend supported by SQLAlchemy can be used. I've tested with Postgres and MySQL.
* Computes a confidence score in the range [0, 1] and the relative offset (which part of the matched song an input song corresponds to) to help evaluate/filter matches.
